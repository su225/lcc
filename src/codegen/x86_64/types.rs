use std::fmt::{Display, Formatter};
use crate::codegen::x86_64::register::Register;
use crate::tacky::TackySymbol;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct StackOffset(pub(crate) isize);

#[derive(Debug, Clone, PartialEq)]
pub enum AsmOperand {
    Imm32(i32),
    Reg(Register),
    Pseudo(TackySymbol),
    Stack { offset: StackOffset },
}

impl Display for AsmOperand {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match &self {
            AsmOperand::Imm32(n) => format!("${}", n),
            AsmOperand::Reg(r) => r.to_string(),
            AsmOperand::Pseudo(p) => format!("<<{}>>", p),
            AsmOperand::Stack { offset } => format!("{}(%rbp)", offset.0),
        })
    }
}

#[derive(Debug, PartialEq)]
pub enum AsmInstruction {
    AllocateStack(usize),
    Mov8 { src: AsmOperand, dst: AsmOperand },
    Mov32 { src: AsmOperand, dst: AsmOperand },
    Neg32 { op: AsmOperand },
    Not32 { op: AsmOperand },
    Add32 { src: AsmOperand, dst: AsmOperand },
    Sub32 { src: AsmOperand, dst: AsmOperand },
    IMul32 { src: AsmOperand, dst: AsmOperand },
    IDiv32 { divisor: AsmOperand },
    And32 { src: AsmOperand, dst: AsmOperand },
    Or32 { src: AsmOperand, dst: AsmOperand },
    Xor32 { src: AsmOperand, dst: AsmOperand },
    Sal32 { src: AsmOperand, dst: AsmOperand },
    Sar32 { src: AsmOperand, dst: AsmOperand },
    Shl32 { src: AsmOperand, dst: AsmOperand },
    Shr32 { src: AsmOperand, dst: AsmOperand },
    SignExtendTo64, // Sign extend %eax to 64-bit into %edx
    Ret,
}

#[derive(Debug, PartialEq)]
pub struct AsmFunction {
    pub name: TackySymbol,
    pub instructions: Vec<AsmInstruction>,
}

#[derive(Debug, PartialEq)]
pub struct AsmProgram {
    pub functions: Vec<AsmFunction>,
}