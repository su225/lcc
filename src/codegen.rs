use std::fmt::{Display, Formatter};
use std::num::ParseIntError;

use thiserror::Error;

use crate::parser::{Expression, ExpressionKind, FunctionDefinition, ProgramDefinition, Statement, StatementKind};

#[derive(Debug)]
pub enum Register {
    RAX,
    RBX,
    RCX,
    RDX,
}

impl Display for Register {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "%{}", match &self {
            Register::RAX => "rax",
            Register::RBX => "rbx",
            Register::RCX => "rcx",
            Register::RDX => "rdx",
        })
    }
}

#[derive(Debug)]
pub enum Operand {
    Imm(i64),
    Register(Register),
}

impl Display for Operand {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match &self {
            Operand::Imm(n) => format!("${}", n),
            Operand::Register(r) => r.to_string(),
        })
    }
}

#[derive(Debug)]
pub enum AsmInstruction {
    Mov { src: Operand, dst: Operand },
    Ret,
}

#[derive(Debug)]
pub struct AsmFunction<'a> {
    pub name: &'a str,
    pub instructions: Vec<AsmInstruction>,
}

#[derive(Debug)]
pub struct AsmProgram<'a> {
    pub functions: Vec<AsmFunction<'a>>,
}

#[derive(Error, Debug)]
pub enum CodegenError {
    #[error(transparent)]
    IntImmediateParseError(#[from] ParseIntError),
}

pub fn generate_assembly<'a>(p: ProgramDefinition<'a>) -> Result<AsmProgram<'a>, CodegenError> {
    let mut asm_functions: Vec<AsmFunction<'a>> = vec![];
    for func in p.functions.iter() {
        let asm = generate_function_assembly(func)?;
        asm_functions.push(asm)
    }
    Ok(AsmProgram { functions: asm_functions })
}

fn generate_function_assembly<'a>(f: &FunctionDefinition<'a>) -> Result<AsmFunction<'a>, CodegenError> {
    let mut body_statements: Vec<AsmInstruction> = vec![];
    for stmt in f.body.iter() {
        let instrs_for_statement = generate_statement_assembly(stmt)?;
        body_statements.extend(instrs_for_statement);
    }
    Ok(AsmFunction {
        name: f.name.name,
        instructions: body_statements,
    })
}

fn generate_statement_assembly(s: &Statement<'_>) -> Result<Vec<AsmInstruction>, CodegenError> {
    match &s.kind {
        StatementKind::Return(expr) => {
            let (mut expr_instrs, res_operand) = generate_expression_assembly(&expr)?;
            expr_instrs.push(AsmInstruction::Mov {
                src: res_operand,
                dst: Operand::Register(Register::RAX),
            });
            expr_instrs.push(AsmInstruction::Ret);
            Ok(expr_instrs)
        }
    }
}

fn generate_expression_assembly(e: &Expression<'_>) -> Result<(Vec<AsmInstruction>, Operand), CodegenError> {
    match e.kind {
        ExpressionKind::IntConstant(num, radix) => {
            let n = i64::from_str_radix(num, radix.value())?;
            Ok((
                vec![
                    AsmInstruction::Mov {
                        src: Operand::Imm(n),
                        dst: Operand::Register(Register::RAX),
                    }
                ],
                Operand::Register(Register::RAX),
            ))
        }
    }
}