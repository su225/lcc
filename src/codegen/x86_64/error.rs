use std::num::ParseIntError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CodegenError {
    #[error(transparent)]
    IntImmediateParseError(#[from] ParseIntError),
}