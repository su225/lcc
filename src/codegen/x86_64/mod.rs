pub mod codegen;
pub mod asmgen;
pub mod register;
pub mod types;

pub use codegen::*;
pub use asmgen::*;