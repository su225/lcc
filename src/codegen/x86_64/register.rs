use std::fmt::{Display, Formatter};
use crate::codegen::x86_64::register::Register::*;

#[derive(Debug, Clone, PartialEq)]
pub enum Register {
    EAX, EDX, R10D, R11D,

    RSP, RBP,

    R10B, AL, CL,
}

impl Display for Register {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "%{}", match &self {
            EAX => "eax",
            EDX => "edx",
            R10D => "r10d",
            R11D => "r11d",

            RSP => "rsp",
            RBP => "rbp",

            R10B => "r10b",
            CL => "cl",
            AL => "al",
        })
    }
}