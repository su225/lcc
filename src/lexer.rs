use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::iter::Peekable;
use std::str::Chars;

use once_cell::sync::Lazy;
use thiserror::Error;

#[derive(Debug, Eq, PartialEq, Clone, Hash)]
pub enum KeywordIdentifier {
    TypeInt,
    TypeVoid,
    Return,
}

type Radix = u8;

const DECIMAL: Radix = 10;
const OCTAL: Radix = 8;
const HEXADECIMAL: Radix = 16;
const BINARY: Radix = 2;

#[derive(Debug, Eq, PartialEq)]
pub enum TokenType<'a> {
    Keyword(KeywordIdentifier),
    OpenParentheses,
    CloseParentheses,
    OpenBrace,
    CloseBrace,
    Semicolon,
    Identifier(&'a str),
    IntConstant(&'a str, Radix),
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
struct Location {
    line: usize,
    column: usize,
}

impl Location {
    fn advance_line(&mut self) {
        self.line += 1;
        self.column = 1;
    }

    fn advance_tab(&mut self) {
        self.column = ((self.column + 7) / 8) * 8;
    }

    fn advance(&mut self) {
        self.column += 1;
    }
}

#[derive(Error, Debug, PartialEq)]
pub enum LexerError {
    #[error("{location:?}: unexpected character")]
    UnexpectedCharacter { location: Location },

    #[error("{location:?}: unknown radix representation of integer")]
    UnknownRadixRepresentation { location: Location },

    #[error("{location:?}: invalid digit {digit:?} for radix {radix:?}")]
    InvalidDigitForRadix { location: Location, digit: char, radix: Radix },

    #[error("{location:?}: invalid character {ch:?} for identifier")]
    InvalidIdentifierCharacter { location: Location, ch: char },
}

type LexerResult<T> = Result<T, LexerError>;

static KEYWORDS: Lazy<HashMap<&'static str, KeywordIdentifier>> = Lazy::new(|| {
    HashMap::from([
        ("int", KeywordIdentifier::TypeInt),
        ("void", KeywordIdentifier::TypeVoid),

        ("return", KeywordIdentifier::Return),
    ])
});

static KEYWORD_STRINGS: Lazy<HashMap<KeywordIdentifier, &'static str>> = Lazy::new(|| {
    HashMap::from([
        (KeywordIdentifier::TypeInt, "int"),
        (KeywordIdentifier::TypeVoid, "void"),

        (KeywordIdentifier::Return, "return"),
    ])
});

impl<'a> Display for TokenType<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match &self {
            TokenType::Keyword(k) => f.write_str(KEYWORD_STRINGS.get(k).unwrap()),
            TokenType::OpenParentheses => f.write_str("("),
            TokenType::CloseParentheses => f.write_str(")"),
            TokenType::OpenBrace => f.write_str("{"),
            TokenType::CloseBrace => f.write_str("}"),
            TokenType::Semicolon => f.write_str(";"),
            TokenType::Identifier(x) => f.write_fmt(format_args!("identifier:{}", x)),
            TokenType::IntConstant(x, radix) => f.write_fmt(format_args!("int:[{}, radix:{}]", x, radix)),
        }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Token<'a> {
    token_type: TokenType<'a>,
    location: Location,
}

pub struct Lexer<'a> {
    input: &'a str,
    char_stream: Peekable<Chars<'a>>,
    cur_stream_pos: usize,
    cur_location: Location,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input,
            char_stream: input.chars().peekable(),
            cur_stream_pos: 0,
            cur_location: Location { line: 1, column: 1 },
        }
    }

    fn next_char(&mut self) -> Option<char> {
        let ch = self.char_stream.next();
        if ch.is_none() {
            return None;
        }
        let c = ch.unwrap();
        self.cur_stream_pos += c.len_utf8();
        match c {
            '\n' => self.cur_location.advance_line(),
            '\t' => self.cur_location.advance_tab(),
            _ => self.cur_location.advance(),
        }
        return ch;
    }

    fn tokenize_single_char(&mut self, token_type: TokenType<'a>) -> Token<'a>  {
        let token = Token {
            token_type,
            location: self.cur_location.clone(),
        };
        self.next_char();
        token
    }

    fn tokenize_integer_constant(&mut self) -> Result<Token<'a>, LexerError> {
        let start_loc = self.cur_location.clone();
        let start_pos = self.cur_stream_pos;
        let first_digit = self.next_char().unwrap();
        let mut expected_radix = OCTAL;
        if first_digit != '0' {
            expected_radix = DECIMAL;
        } else {
            let second_digit = self.next_char();
            if second_digit.is_none() {
                return Ok(Token { location: start_loc, token_type: TokenType::IntConstant("0", DECIMAL) });
            }
            let digit2 = second_digit.unwrap();
            if digit2 == 'x' || digit2 == 'X' {
                expected_radix = HEXADECIMAL;
            } else if digit2 == 'b' || digit2 == 'B' {
                expected_radix = BINARY;
            } else if !digit2.is_digit(OCTAL as u32) {
                // If it is neither 0x, 0X, 0b, 0B or 0[Octal digit]
                // then it is an invalid number. Hence, it is an error.
                return Ok(Token { location: start_loc, token_type: TokenType::IntConstant("0", DECIMAL) });
            }
        }
        // Once we have determined the radix, we iterate through the digits as long
        // as we find a valid stop point: semicolon, whitespace (space, new-line, tab), end.
        // If we stop at characters that are not valid digits, then we throw an error. We
        // don't support digit grouping yet to keep the lexer simple.
        loop {
            let loc = self.cur_location.clone();
            let next = self.char_stream.peek();
            if next.is_none() {
                break;
            }
            let n = next.unwrap().clone();
            if !n.is_digit(expected_radix as u32) {
                break;
            }
            self.next_char();
        }
        return Ok(Token {
            location: start_loc,
            token_type: TokenType::IntConstant(&self.input[start_pos..self.cur_stream_pos], expected_radix),
        })
    }

    fn tokenize_identifier_or_keyword(&mut self) -> Result<Token<'a>, LexerError> {
        let start_loc = self.cur_location;
        let start_pos = self.cur_stream_pos;
        let first_char = self.next_char().unwrap();
        if first_char != '_' && !first_char.is_alphabetic() {
            return Err(LexerError::InvalidIdentifierCharacter{ location: start_loc, ch: first_char });
        }
        loop {
            let cur_loc = self.cur_location;
            let ch = self.char_stream.peek();
            if ch.is_none() {
                break;
            }
            let c = ch.unwrap().clone();
            if !c.is_alphanumeric() && c != '_' {
                break;
            }
            self.next_char();
        }
        let word = &self.input[start_pos..self.cur_stream_pos];
        match KEYWORDS.get(word) {
            None => Ok(Token{
                location: start_loc,
                token_type: TokenType::Identifier(word),
            }),
            Some(k) => Ok(Token{
                location: start_loc,
                token_type: TokenType::Keyword(k.clone()),
            })
        }
    }

    fn next_token(&mut self) -> Result<Option<Token<'a>>, LexerError> {
        loop {
            let cur = self.char_stream.peek();
            if cur.is_none() {
                // We have reached the end of the stream. Hence, we cannot
                // tokenize anymore. So we just return None
                return Ok(None);
            }
            let cur_char = cur.unwrap();
            match *cur_char {
                ';' => {
                    let token = self.tokenize_single_char(TokenType::Semicolon);
                    return Ok(Some(token));
                },
                '(' => {
                    let token = self.tokenize_single_char(TokenType::OpenParentheses);
                    return Ok(Some(token));
                },
                ')' => {
                    let token = self.tokenize_single_char(TokenType::CloseParentheses);
                    return Ok(Some(token));
                },
                '{' => {
                    let token = self.tokenize_single_char(TokenType::OpenBrace);
                    return Ok(Some(token));
                },
                '}' => {
                    let token = self.tokenize_single_char(TokenType::CloseBrace);
                    return Ok(Some(token));
                },
                '0'..='9' => {
                    let token = self.tokenize_integer_constant()?;
                    return Ok(Some(token));
                },
                '\n' | '\t' | ' ' => {
                    self.next_char();
                    continue;
                },
                _ => {
                    let token = self.tokenize_identifier_or_keyword()?;
                    return Ok(Some(token));
                },
            }
        }
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Result<Token<'a>, LexerError>;

    fn next(&mut self) -> Option<Self::Item> {
        let next_tok = self.next_token();
        match next_tok {
            Ok(Some(tok)) => Some(Ok(tok)),
            Ok(None) => None,
            Err(e) => Some(Err(e)),
        }
    }
}

#[cfg(test)]
mod test {
    use std::collections::HashMap;

    use crate::lexer::{BINARY, DECIMAL, HEXADECIMAL, KEYWORDS, Lexer, LexerError, LexerResult, Location, OCTAL, Token, TokenType};
    use crate::lexer::TokenType::Keyword;

    #[test]
    fn test_tokenizing_open_and_close_parentheses() {
        let source = "()";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 2 }},
        ]));
    }

    #[test]
    fn test_tokenizing_open_and_close_braces() {
        let source = "{}";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenBrace, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseBrace, location: Location { line: 1, column: 2 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_newlines() {
        let source = "(\n\n)";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 3, column: 1 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_tabs() {
        let source = "(\t)";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 8 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_whitespaces() {
        let source = "(   )";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 5 }},
        ]));
    }

    #[test]
    fn test_tokenizing_semicolon() {
        let source = ";";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::Semicolon, location: Location { line: 1, column: 1 }},
        ]));
    }

    #[test]
    fn test_tokenizing_valid_identifiers() {
        let identifiers = vec![
            "abcde",
            "abcde123",
            "hello_world_123",
            "helloWorld123",
            "_abcde",
            "_123",
            "_123_456",
            "café",
            "αριθμός",
            "число",
            "数字",
        ];
        for src in identifiers.into_iter() {
            let lexer = Lexer::new(src);
            let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
            let expected_tokens: LexerResult<Vec<Token>> = Ok(vec![
                Token { token_type: TokenType::Identifier(src), location: Location { line: 1, column: 1 }},
            ]);
            assert_eq!(tokens, expected_tokens,
                       "lexing identifier {}: expected: {:?}, actual:{:?}",
                        src, expected_tokens, tokens);
        }
    }

    #[test]
    fn test_tokenizing_keywords() {
        for (kw, kwid) in KEYWORDS.iter() {
            let lexer = Lexer::new(kw);
            let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
            let expected_tokens: LexerResult<Vec<Token>> = Ok(vec![
                Token { token_type: Keyword(kwid.clone()), location: Location { line: 1, column: 1 } },
            ]);
            assert_eq!(tokens, expected_tokens,
                      "lexing keyword identifier {}: expected: {:?}, actual:{:?}",
                      kw, expected_tokens, tokens);
        }
    }

    #[test]
    fn test_tokenizing_valid_integers() {
        let integers_base10 = vec![
            "1",
            "0",
            "2",
            "3",
            "44",
            "55",
            "66",
            "777",
            "888",
            "9189",
            "189087931798698368761873",
            "0\t\t",
            "0  ",
            "0\n",
        ];

        let integers_hex = vec!["0xdeadbeef",
            "0Xcafebabe",
            "0XDEADBEEF",
            "0xCAFEbabe",
        ];

        let integers_octal = vec![
            "000",
            "01234567",
            "012300",
        ];

        let integers_binary = vec![
            "0b01010010010",
            "0b001001001",
        ];

        let int_tests = HashMap::from([
            (10, integers_base10),
            (16, integers_hex),
            (8, integers_octal),
            (2, integers_binary),
        ]);


        for (base, srcs) in int_tests.into_iter() {
            for src in srcs.into_iter() {
                let lexer = Lexer::new(src);
                let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
                let expected_tokens = Ok(vec![
                    Token { token_type: TokenType::IntConstant(src.trim(), base), location: Location { line: 1, column: 1 }},
                ]);
                assert_eq!(tokens, expected_tokens,
                           "lexing identifier {}: expected: {:?}, actual:{:?}",
                           src, expected_tokens, tokens);
            }
        }
    }

    #[test]
    fn test_bad_identifier_error() {
        let src = "abcd@";
        let lexer = Lexer::new(src);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        let expected = Err(LexerError::InvalidIdentifierCharacter {
            location: Location {line: 1, column: 5},
            ch: '@',
        });
        assert_eq!(tokens, expected);
    }
}