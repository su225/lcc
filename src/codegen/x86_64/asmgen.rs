use std::io;
use std::io::Write;
use crate::codegen::x86_64::*;
use crate::codegen::x86_64::AsmInstruction::*;

pub fn emit_assembly<W: Write>(asm_code: AsmProgram, mut w: W) -> io::Result<()> {
    for f in asm_code.functions.into_iter() {
        emit_function(f, &mut w)?;
    }
    Ok(())
}

fn emit_function<W: Write>(f: AsmFunction, w: &mut W) -> io::Result<()> {
    w.write_fmt(format_args!("    .globl _{function_name}\n", function_name = f.name))?;
    w.write_fmt(format_args!("_{function_name}:\n", function_name = f.name))?;
    emit_function_prologue(w)?;
    for instr in f.instructions {
        emit_instruction(&instr, w)?;
    }
    Ok(())
}

macro_rules! emit_instruction {
    ($w:expr, $($arg:tt)*) => {
        writeln!($w, "    {}", format!($($arg)*))
    };
}

fn emit_instruction<W: Write>(instr: &AsmInstruction, w: &mut W) -> io::Result<()> {
    match instr {
        Mov8 { src, dst } => emit_instruction!(w, "movb {src}, {dst}")?,
        Mov32 { src, dst } => emit_instruction!(w, "movl {src}, {dst}")?,
        Ret => {
            emit_function_epilogue(w)?;
            emit_instruction!(w, "ret")?
        },
        Not32 { op } => emit_instruction!(w, "notl {op}")?,
        Neg32 { op } => emit_instruction!(w, "negl {op}")?,
        AllocateStack(stack_size) => emit_instruction!(w, "subq ${stack_size}, %rsp")?,
        Add32 { src, dst } => emit_instruction!(w, "addl {src}, {dst}")?,
        Sub32 { src, dst } => emit_instruction!(w, "subl {src}, {dst}")?,
        IMul32 { src, dst } => emit_instruction!(w, "imull {src}, {dst}")?,
        IDiv32 { divisor } => emit_instruction!(w, "idivl {divisor}")?,
        SignExtendTo64 => emit_instruction!(w, "cdq")?,

        And32 { src, dst } => emit_instruction!(w, "andl {src}, {dst}")?,
        Or32 { src, dst } => emit_instruction!(w, "orl {src}, {dst}")?,
        Xor32 { src, dst } => emit_instruction!(w, "xorl {src}, {dst}")?,
        Shl32 { src, dst } => emit_instruction!(w, "shll {src}, {dst}")?,
        Shr32 { src, dst } => emit_instruction!(w, "shrl {src}, {dst}")?,
        Sal32 { src, dst } => emit_instruction!(w, "sall {src}, {dst}")?,
        Sar32 { src, dst } => emit_instruction!(w, "sarl {src}, {dst}")?,

        Cmp32 { op1, op2 } => emit_instruction!(w, "cmpl {op1}, {op2}")?,
        Jmp { target } => emit_instruction!(w, "jmp {target}")?,
        JmpConditional { condition_code, target_if_true } =>
            emit_instruction!(w, "j{condition_code} {target_if_true}")?,
        SetCondition { condition_code, dst } =>
            emit_instruction!(w, "set{condition_code} {dst}")?,
        Label(lbl) => writeln!(w, "{}", format!("{lbl}:"))?,
    };
    Ok(())
}

fn emit_function_prologue<W: Write>(w: &mut W) -> io::Result<()> {
    emit_instruction!(w, "pushq %rbp")?;
    emit_instruction!(w, "movq  %rsp, %rbp")?;
    Ok(())
}

fn emit_function_epilogue<W: Write>(w: &mut W) -> io::Result<()> {
    emit_instruction!(w, "movq %rbp, %rsp")?;
    emit_instruction!(w, "popq %rbp")?;
    Ok(())
}

#[cfg(test)]
mod test {
    use std::fs;
    use std::io::Cursor;
    use std::path::{Path, PathBuf};
    use insta::assert_snapshot;
    use rstest::rstest;
    use crate::codegen::x86_64::{emit_assembly, generate_assembly};
    use crate::lexer::Lexer;
    use crate::parser::Parser;
    use crate::semantic_analysis::identifier_resolution::resolve_program;
    use crate::semantic_analysis::loop_labeling::loop_label_program_definition;
    use crate::tacky::emit;

    #[rstest]
    #[case("multifunc/simple.c")]
    fn test_generation_for_multiple_functions(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("multiple functions", input_path);
    }

    #[rstest]
    #[case("unary/complement.c")]
    #[case("unary/negation.c")]
    #[case("unary/not.c")]
    fn test_generation_for_unary_operators(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("unary operators", input_path)
    }

    #[rstest]
    #[case("binary/arithmetic_add.c")]
    #[case("binary/arithmetic_add_leftassoc.c")]
    #[case("binary/arithmetic_precedence.c")]
    #[case("binary/arithmetic_precedence_override.c")]
    #[case("binary/arithmetic_multiplication.c")]
    #[case("binary/arithmetic_division.c")]
    #[case("binary/arithmetic_modulo.c")]
    fn test_generation_for_arithmetic_binary_operators(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("arithmetic binary operators", input_path)
    }

    #[rstest]
    #[case("binary/bitwise_and.c")]
    #[case("binary/bitwise_or.c")]
    #[case("binary/bitwise_xor.c")]
    #[case("binary/bitwise_left_shift.c")]
    #[case("binary/bitwise_right_shift.c")]
    fn test_generation_for_bitwise_binary_operators(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("arithmetic bitwise binary operators", input_path)
    }

    #[rstest]
    #[case("binary/relational_eq.c")]
    fn test_generation_for_relational_binary_operators(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("arithmetic relational binary operators", input_path)
    }

    #[rstest]
    #[case("binary/logical_and_false.c")]
    fn test_generation_for_logical_and_false(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("logical operators with short-circuiting", input_path)
    }

    #[rstest]
    #[case("localvars/simple.c")]
    #[case("localvars/declaration_and_assign.c")]
    #[case("localvars/nested_scopes.c")]
    #[case("localvars/expression_with_var.c")]
    fn test_generation_for_local_vars(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("local variables", input_path)
    }

    #[rstest]
    #[case("conditional/if.c")]
    #[case("conditional/if_else.c")]
    #[case("conditional/if_else_if.c")]
    #[case("conditional/dangling_if.c")]
    #[case("conditional/ternary.c")]
    fn test_generation_for_conditional_statements(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("conditional statements", input_path)
    }

    #[rstest]
    #[case("loops/simple_for.c")]
    #[case("loops/for_init_decl.c")]
    #[case("loops/for_init_decl_shadow.c")]
    #[case("loops/for_noinit.c")]
    #[case("loops/for_all_empty.c")]
    #[case("loops/for_with_break.c")]
    #[case("loops/for_with_continue.c")]
    #[case("loops/nested_for.c")]
    #[case("loops/nested_for_with_break.c")]
    #[case("loops/nested_for_with_continue.c")]
    fn test_generation_for_for_loop(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("loops: for", input_path)
    }

    #[rstest]
    #[case("loops/simple_while.c")]
    #[case("loops/while_with_break.c")]
    #[case("loops/while_with_continue.c")]
    fn test_generation_for_while_loop(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("loops: while", input_path)
    }

    #[rstest]
    #[case("loops/simple_dowhile.c")]
    #[case("loops/dowhile_with_break.c")]
    #[case("loops/dowhile_with_continue.c")]
    fn test_generation_for_dowhile_loop(#[case] input_path: &str) {
        run_asm_emit_snapshot_test("loops: do-while", input_path)
    }

    fn run_asm_emit_snapshot_test(suite_description: &str, src_file: &str) {
        let base_dir = file!();
        let src_path = Path::new(base_dir).parent().unwrap().join("input").join(src_file);
        let source = fs::read_to_string(src_path.clone());
        assert!(source.is_ok(), "failed to read {:?}", src_path);

        let src = source.unwrap();
        let lexer = Lexer::new(&src);
        let mut parser = Parser::new(lexer);
        let ast = parser.parse().expect("parsing failed");
        let id_resolved_ast = resolve_program(ast).expect("identifier resolution failed");
        let loop_labeled_ast = loop_label_program_definition(id_resolved_ast).expect("loop labeling failed");
        let tacky = emit(&loop_labeled_ast).expect("tacky gen failed");
        let asm = generate_assembly(tacky).expect("asm gen failed");

        let mut buffer = Cursor::new(Vec::new());
        emit_assembly(asm, &mut buffer).expect("asm emit failed");
        let asm_output = String::from_utf8(buffer.into_inner()).expect("invalid utf-8");

        let (out_dir, snapshot_file) = output_path_parts(src_file);
        insta::with_settings!({
            sort_maps => true,
            prepend_module_to_snapshot => false,
            description => suite_description,
            snapshot_path => out_dir,
            info => &format!("{}", src_file),
        }, {
            assert_snapshot!(snapshot_file, asm_output);
        });
    }

    fn output_path_parts(src_file: &str) -> (PathBuf, String) {
        let input_path = Path::new(src_file);
        let parent = input_path.parent().unwrap_or_else(|| Path::new(""));
        let stem = input_path.file_stem().expect("No file stem").to_string_lossy();
        let output_dir = Path::new("output").join(parent);
        let output_file = format!("{}.s", stem);
        (output_dir, output_file)
    }
}