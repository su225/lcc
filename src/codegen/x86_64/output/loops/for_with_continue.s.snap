---
source: src/codegen/x86_64/asmgen.rs
description: "loops: for"
expression: asm_output
info: loops/for_with_continue.c
---
    .globl _main
_main:
    pushq %rbp
    movq  %rsp, %rbp
    subq $72, %rsp
    movl $0, -8(%rbp)
    movl $0, -16(%rbp)
    movl $0, -8(%rbp)
.loop.0.start:
    cmpl $10, -8(%rbp)
    movl $0, -24(%rbp)
    setl -24(%rbp)
    movl -24(%rbp), %r10d
    movl %r10d, -32(%rbp)
    cmpl $0, -32(%rbp)
    je .loop.0.break
    movl -8(%rbp), %eax
    cdq
    movl $2, %r10d
    idivl %r10d
    movl %edx, -40(%rbp)
    cmpl $0, -40(%rbp)
    movl $0, -48(%rbp)
    sete -48(%rbp)
    movl -48(%rbp), %r10d
    movl %r10d, -56(%rbp)
    cmpl $0, -56(%rbp)
    je _L.if_end.0
    jmp .loop.0.continue
_L.if_end.0:
    movl -16(%rbp), %r10d
    movl %r10d, -64(%rbp)
    movl -8(%rbp), %r10d
    addl %r10d, -64(%rbp)
    movl -64(%rbp), %r10d
    movl %r10d, -16(%rbp)
.loop.0.continue:
    movl -8(%rbp), %r10d
    movl %r10d, -72(%rbp)
    movl -8(%rbp), %r10d
    movl %r10d, -8(%rbp)
    addl $1, -8(%rbp)
    jmp .loop.0.start
.loop.0.break:
    movl -16(%rbp), %eax
    movq %rbp, %rsp
    popq %rbp
    ret
